### ISSUE TITLE
A strange bug in fonts generation, since 1.8.0

### ISSUE THREAD

**Issue by serhii-k** [libgdx/libgdx#3791]
Hi guys,

I have created a simple project that shows this bug.

Please, clone it:
git clone https://github.com/serhii-k/test.git

It contains the ttf-file that contains only one character 'm', for the simplicity of this test.

When I generate a font, specifying:
parameter.characters = "m";

Then
font.draw(batch, "m", 300, 300);
displays this result:

![test1](https://cloud.githubusercontent.com/assets/10563821/12652654/778483f2-c5f4-11e5-8e7d-f7f507cfcbe9.png)

A workaround is to specify the characters parameter as "mm":
parameter.characters = "mm";

Then, the font is renderred correctly:

![test2](https://cloud.githubusercontent.com/assets/10563821/12652680/938fa536-c5f4-11e5-97d3-3167541ef7c4.png)

This issue appeared in 1.8.0.

Kind regards,
Serhii

**Comment by serhii-k** [comment]
The described workaround doesn't work well on every platform. E.g. on some iOS simulators that workaround doesn't give consistent results. On some, 3 same characters are required, on some, 2 characters are too much... You can't guarantee which characters need this workaround. Etc.

At this stage, instead of using FreeTypeFontGenerator.generateFont() I'm using FreeTypeFontGenerator.generateGlyphAndBitmap() for each separate character (I'm using this technique to generate app icons from a ttf-font). Then I'm packing pixmaps obtained from bitmaps with PixmapPacker to get a TextureAtlas with icons.

It looks that the generateGlyphAndBitmap method works consistently on different platforms.

**Comment by Darkyenus** [comment]
Creating example code as described [here](https://github.com/libgdx/libgdx/wiki/Getting-Help#executable-example-code) instead of whole repo will make it easier for others to help.

That said, looking at your code: does this happen with smaller glyph sizes? 150 is pretty big.

**Comment by intrigus** [comment]
Darkyenus I can confirm that this bug exists. 

This bug does not occur, when FreeType is in incremental mode.

Here is my reproduction case:
````
public class FreeTypeOneCharTest extends GdxTest {

	SpriteBatch batch;
	BitmapFont font;
	FreeTypeFontGenerator generator;
	FreeTypeBitmapFontData data;

	public void create () {
		batch = new SpriteBatch();

		generator = new FreeTypeFontGenerator(Gdx.files.internal("data/arial.ttf"));

		FreeTypeFontParameter param = new FreeTypeFontParameter();
		param.incremental = false; //Switch to true and it will magically work
		param.size = 20;
		param.color = Color.GREEN;
		param.shadowColor = Color.YELLOW;
		param.shadowOffsetX = 6;
		param.shadowOffsetY = 6;
		param.characters = "s";

		data = new FreeTypeBitmapFontData();

		font = generator.generateFont(param, data);
	}

	public void render () {
		Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT);

		batch.begin();
		font.draw(batch, "s", 10, 300);
		batch.end();
	}

	public void resize (int width, int height) {
		batch.getProjectionMatrix().setToOrtho2D(0, 0, width, height);
	}

	@Override
	public void dispose () {
		generator.dispose();
		font.dispose();
		batch.dispose();
	}

}
````

**Comment by Darkyenus** [comment]
Try viewing generated textures to check if the problem is in the rasterization or in rendering. [This may help with that.](https://gist.github.com/Darkyenus/87f12ef9df05762510b6) Just put it in a `if(Gdx.input.justTouched())` block in the render method.

**Comment by intrigus** [comment]
Thanks for the snippet, will try debugging this issue tomorrow.

**Comment by serhii-k** [comment]
Oh wow, @intrigus I can confirm that setting incremental to true resolves the issue in my example too.

I'm sorry @Darkyenus I sould have definitely added a simple barebones class.

In my code I was taking textures from the generated font with icons and rendered them. So I suppose the problem is in the rasterization. I will check this assumption a little later.

**Comment by intrigus** [comment]
I can confirm that FreeType itself renders the glyphs correctly. (The pixmaps returned by https://github.com/libgdx/libgdx/blob/master/extensions/gdx-freetype/src/com/badlogic/gdx/graphics/g2d/freetype/FreeType.java#L607 are the same for `incremental=true` and `false`)
So the problem seems to be in the packaging of the glyphs to the texture, will look into this later.

**Comment by serhii-k** [comment]
> So the problem seems to be in the packaging of the glyphs to the texture, will look into this later.

Yes, this is absolutely true. In this post I will provide some observed examples.

Here's a barebones version of my code. Please, download the test-font here:
https://github.com/serhii-k/test/blob/master/android/assets/test-font.ttf

```
import com.badlogic.gdx.ApplicationAdapter;
import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.backends.lwjgl.LwjglApplication;
import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.GL20;
import com.badlogic.gdx.graphics.g2d.BitmapFont;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.badlogic.gdx.graphics.g2d.freetype.FreeTypeFontGenerator;
import com.badlogic.gdx.utils.Array;

public class TestFont extends ApplicationAdapter {

    SpriteBatch batch;
    BitmapFont font;

    Array<TextureRegion> regions;

    @Override
    public void create() {
        batch = new SpriteBatch();

        FreeTypeFontGenerator generator = new FreeTypeFontGenerator(Gdx.files.internal("test-font.ttf"));
        FreeTypeFontGenerator.FreeTypeFontParameter parameter = new FreeTypeFontGenerator.FreeTypeFontParameter();
        parameter.characters = "mm";
        parameter.size = 150;
        parameter.gamma = 0.5f;
        font = generator.generateFont(parameter);
        generator.dispose();

        regions = font.getRegions();
        Gdx.app.log("getRegions.size", regions.size + "");
    }

    @Override
    public void render() {
        Gdx.gl.glClearColor(0.2f, 0.2f, 0.2f, 1f);
        Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT);
        batch.begin();

        batch.setColor(Color.WHITE);
        font.draw(batch, "m", 100, 450);

        batch.setColor(Color.BLUE);
        batch.draw(regions.get(0), 10, 10, regions.get(0).getRegionWidth(), regions.get(0).getRegionHeight());

        if (regions.size > 1) {
            batch.setColor(Color.RED);
            batch.draw(regions.get(1), 10 + regions.get(0).getRegionWidth() + 10, 10,
                    regions.get(1).getRegionWidth(), regions.get(1).getRegionHeight());
        }

        batch.end();
    }

    public static void main (String[] args) throws Exception {
        new LwjglApplication(new TestFont());
    }
}
```

if `parameter.characters = "m";` only one 256x256 region is generated. But, look, it has some awkward square... Strange, ha? The actual character is placed below this square and partially is outside the region...

![m](https://cloud.githubusercontent.com/assets/10563821/13350008/296fdede-dc86-11e5-8bed-39474ab7e91a.png)

If `parameter.characters = "mm";` two 256x256 regions are generated. The first one is same as in the first case, the second one has a correctly placed character.

![mm](https://cloud.githubusercontent.com/assets/10563821/13350052/55e6e688-dc86-11e5-88de-44f4132331fc.png)

So, the issue is really in the packaging of the glyphs to the texture as @intrigus has noted. In the example, the strange square appears.

If the font size is lower, let's say 32 pixels instead of 150, the issue remains the same. Only the generated regions are smaller.

**Comment by Darkyenus** [comment]
Incremental mode uses different pixmap packer implementation, so that explains it.

**Comment by mrjohs** [comment]
Hi. Any progress on this? I have been having this issue since upgrading to 1.9.4 from 1.6.5, when using the ttf from font awesome.

Some observations:

- Sometimes the icons are rendered correctly on the desktop version and not on the android version. The android version has a higher dpi so the icons will have more pixels.
- Sometimes when I add just 2 to the parameter.size, they render correctly.

**Comment by dylan9** [comment]
Hi. Has there been any progress on this? I'm packing my images using PixmapPackers SkylineStrategy and its giving me this problem. I can't use GuillotineStrategy because the packing isn't tight enough, making images overflow.
